third_party/nvfuser/csrc/lower_loop_rotation.cpp

```diff
@@ -0,0 +1,386 @@
+          }
+          if (GpuLower::current()->doubleBufferInfo().getDoubleBufferLoop(
+                  tv, {loop_}) != loop_) {
+            result_ = false;
+            return;
+          }
+        }
+        IrVisitor::handle(expr);
+      }
+
+     private:
+      bool result_ = true;
+      kir::ForLoop* const loop_ = nullptr;
+    } is_double_buffer_load(fl);
+
+    bool seen_unselected = false;
+    for (auto expr : fl->body().exprs()) {
+      if (selection.count(expr) > 0) {
+        if (seen_unselected) {
+          return false;
+        }
+        continue;
+      }
+      if (is_double_buffer_load(expr)) {
+        continue;
+      }
+      seen_unselected = true;
+    }
+    return true;
+  }
+
+  // Do the rotation. If I have loop:
+  //   for (int i = 0; i < n; i++) {
+  //     selected1(i);
+  //     selected2(i);
+  //     unselected3(i);
+  //     unselected4(i);
+  //   }
+  // Then this function should generate
+  //   // prologue
+  //   if (0 < n) {
+  //     // trivial loop
+  //     for (int i = 0) {
+  //       selected1(i);
+  //       selected2(i);
+  //     }
+  //   }
+  //   // main loop
+  //   for (int i = 0; i < n; i++) {
+  //     unselected3(i);
+  //     unselected4(i);
+  //     if (i + 1 < n) {
+  //       selected1(i + 1);
+  //       selected2(i + 1);
+  //     }
+  //   }
+  // Currently, because all out-of-bound access are already covered other
+  // predicates, so we are actually generating
+  //   // prologue
+  //   // trivial loop
+  //   for (int i = 0) {
+  //     selected1(i);
+  //     selected2(i);
+  //   }
+  //   // main loop
+  //   for (int i = 0; i < n; i++) {
+  //     unselected3(i);
+  //     unselected4(i);
+  //     if (true) {
+  //       selected1(i + 1);
+  //       selected2(i + 1);
+  //     }
+  //   }
+  void rotate(kir::ForLoop* fl, std::unordered_set<Statement*> selection) {
+    if (isDebugDumpEnabled(DebugDumpOption::LoopRotation)) {
+      std::cout << "Rotating loop:" << std::endl << fl->toString() << std::endl;
+    }
+    // prologue
+    // Currently, all existing predicates should be able to cover the condition
+    // of start < end, so no predicate here. In the future, if we decide that
+    // we need to predicate this, then we should add an kir::IfThenElse here.
+    auto prologue = IrBuilder::create<kir::ForLoop>(
+        fl->iter_domain(), fl->start(), fl->doubleBufferLoopStage());
+    std::vector<Expr*> lifted_alloc;
+    for (auto expr : fl->body().exprs()) {
+      if (selection.count(expr) == 0) {
+        continue;
+      }
+      if (expr->isA<kir::Allocate>()) {
+        lifted_alloc.push_back(expr);
+        continue;
+      }
+      prologue->body().push_back(recursivelyClone(expr));
+    }
+    if (prologue->empty()) {
+      if (isDebugDumpEnabled(DebugDumpOption::LoopRotation)) {
+        std::cout << "Nothing to do." << std::endl;
+      }
+      return;
+    }
+    for (auto expr : lifted_alloc) {
+      registerInsertBefore(fl, expr);
```

Comment:
Before this, each `expr` is in the body of the `fl` loop. Does this registration here also remove the expression from the `fl` loop body? Otherwise, doesn't it mean it exists in both before `fl` and its body?

Reply:
Ah, you're replacing `fl` with a new loop, so this shouldn't cause any problem? Maybe some comment would be helpful to understand.

Reply:
Your understanding is correct. Comment added.